module/zfs/dsl_deadlist.c

```diff
@@ -131,6 +146,23 @@ dsl_deadlist_load_tree(dsl_deadlist_t *dl)
 	ASSERT(MUTEX_HELD(&dl->dl_lock));
 
 	ASSERT(!dl->dl_oldfmt);
+	if (dl->dl_havecache) {
+		/*
+		 * After loading the tree, the caller may modify the tree,
+		 * e.g. to add or remove nodes, or to make a node no longer
+		 * refer to the empty_bpobj.  These changes would make the
+		 * dl_cache incorrect.  Therefore we discard the cache here,
+		 * so that it can't become incorrect.
+		 */
+		dsl_deadlist_cache_entry_t *dlce;
+		void *cookie = NULL;
+		while ((dlce = avl_destroy_nodes(&dl->dl_cache, &cookie))
+		    != NULL) {
```

Comment:
Aren't you missing a `bpobj_close(&dle->dle_bpobj);` here?

It might be nice to slightly refactor the code so the `dsl_deadlist_discard_tree()` function only dropped the cache and could be used here.  Then add a `dsl_deadlist_destroy_tree()` function which does what the current `dsl_deadlist_discard_tree()` function does.

Reply:
No. `uint64_t dlce_bpobj` is the object number. `bpobj_close()` is called on a `bpobj_t*`.

There's no duplicated code between dsl_deadlist_discard_tree() dsl_deadlist_load_tree.  One operates on `dl_cache`, and the other on `dl_tree`.

Reply:
Ahh, I see what you're saying.